### PR TITLE
feat: Allow providing neither a controlled nor an initial value

### PR DESCRIPTION
**Pull request by amannn**

### CODE REVIEW

File: packages/use-optionally-controlled-state/test/index.test.tsx [hunk 1]
```diff
@@ -102,3 +110,43 @@ it('throws when switching from controlled to uncontrolled mode', () => {
...
+
+// Expected return type: `[boolean, (value: boolean) => void]`
+function UncontrolledWithInitialValue(opts: {
+  controlledValue?: boolean;
+  initialValue: boolean;
+}) {
+  const [value, setValue] = useOptionallyControlledState(opts);
+
+  setValue(true);
+  return value.valueOf();
```
Comment (amannn) [inline packages/use-optionally-controlled-state/test/index.test.tsx:134]: This is an error currently, as `value` can be `undefined` …

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 2]
```diff
@@ -1,17 +1,38 @@
...
+// Uncontrolled with initial value
+export default function useOptionallyControlledState<Value>({
+  controlledValue,
+  initialValue,
+  onChange
+}: {
+  controlledValue?: Value;
+  initialValue: Value;
+  onChange?(value: Value): void;
+}): [Value, (value: Value) => void];
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:35]: I had the idea to use function overloading to be able to specify the pairs of arguments and return types together …

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 3]
```diff
@@ -1,37 +1,29 @@
...
+}: {
+  controlledValue?: ControlledValue;
+  initialValue?: InitialValue;
+  onChange?(value: Value): void;
+}): [Value, (value: Value) => void] {
   const isControlled = controlledValue !== undefined;
   const initialIsControlled = useConstant(() => isControlled);
-  const [stateValue, setStateValue] = useState(initialValue);
+  const [stateValue, setStateValue] = useState<Value | undefined>(
+    initialValue as Value | undefined
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:30]: 🤷‍♂️

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 4]
```diff
@@ -53,8 +45,8 @@ export default function useOptionallyControlledState<Value>({
       if (!isControlled) setStateValue(nextValue);
       if (onChange) onChange(nextValue);
     },
-    [onChange, isControlled]
+    [isControlled, onChange]
   );
 
-  return [value, onValueChange];
+  return [value as Value, onValueChange];
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:58]: 🤷‍♂️

File: packages/use-optionally-controlled-state/test/index.test.tsx [hunk 5]
```diff
@@ -102,3 +110,64 @@ it('throws when switching from controlled to uncontrolled mode', () => {
...
+  initialValue?: boolean;
+}) {
+  // False positive, probably need to upgrade ESLint
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [value, setValue] = useOptionallyControlledState(opts);
+
+  setValue(true);
+
+  // @ts-expect-error This should be an error as the value can be undefined
+  return value.valueOf();
```
Comment (amannn) [inline packages/use-optionally-controlled-state/test/index.test.tsx:163]: This should be an error but isn't unfortunately.

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 6]
```diff
@@ -1,37 +1,36 @@
...
-      controlledValue: Value;
-      initialValue?: Value;
-      onChange?(value: Value): void;
-    };
-
 /**
  * Enables a component state to be either controlled or uncontrolled.
  */
-export default function useOptionallyControlledState<Value>({
+export default function useOptionallyControlledState<
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:7]: TODO:
- [ ] Rename to `useOptionalState` as it delivers the same meaning ("controlled" means there's no intermediate state)

File: packages/use-optionally-controlled-state/test/index.test.tsx (hunk 5 above)
Comment (amannn) [reply packages/use-optionally-controlled-state/test/index.test.tsx:163]: @menelaos Do you have an idea by chance?

Comment (menelaos) [reply packages/use-optionally-controlled-state/test/index.test.tsx:163]: I just tried for quite a bit, but couldn't find a solution 😞 I would also suggest to proceed with the potential `undefined` for now.

File: README.md [hunk 7]
```diff
@@ -5,7 +5,7 @@ A collection of commonly used hooks for React apps.
 **Packages:**
 
 - [`use-promised`](./packages/use-promised)
-- [`use-optionally-controlled-state`](./packages/use-optionally-controlled-state)
+- [`use-optional-state`](./packages/use-optional-state)
```
Comment (menelaos) [inline README.md:8]: 👍

File: packages/use-optional-state/README.md [hunk 8]
```diff
@@ -18,21 +18,21 @@ When implementing a component, it's sometimes hard to choose one or the other si
 
 This hook helps you to support both patterns in your components, increasing flexibility while also ensuring ease of use.
 
-Since the solution can be applied on a per-prop basis, you can even enable this behaviour for multiple props that are orthogonal (e.g. a `<Prompt isOpen inputValue="" />` component).
+Since the solution can be applied on a per-prop basis, you can also enable this behaviour for multiple props that are orthogonal (e.g. a `<Prompt isOpen inputValue="" />` component).
```
Comment (menelaos) [inline packages/use-optional-state/README.md:21]: It seems that the initial excitement has worn off 😉

**Comment by menelaos** [comment]
I think we can also drop [this](https://github.com/amannn/react-hooks/blob/449892538e8c32177cc4c3038b47bd8a9828172c/packages/use-optionally-controlled-state/src/index.tsx#L49) type assertion now (the `!`). The respective comment should probably be updated as well...

File: packages/use-optional-state/README.md (hunk 8 above)
Comment (amannn) [reply packages/use-optional-state/README.md:21]: Haha, right 😆. I found it unnecessarily salesy on a second look 😁

**Comment by amannn** [comment]
> I think we can also drop [this](https://github.com/amannn/react-hooks/blob/449892538e8c32177cc4c3038b47bd8a9828172c/packages/use-optionally-controlled-state/src/index.tsx#L49) type assertion now (the `!`). The respective comment should probably be updated as well...

Great point, thanks!

**Review by menelaos — APPROVED** [review]

**Pull request merged**